janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -546,13 +572,42 @@ private Response performRequest(String method, String path, byte[] requestData)
         return performRequest(new Request(method, path), requestData);
     }
 
+    private Response performRequestWithRetry(Request request) throws IOException {
+        int retryCount = 0;
+        ResponseException lastException;
+        do {
+            if (retryCount > 0) {
+                //Wait before trying again
+                long waitDurationMs = Math.min((long) (retryInitialWaitMs * Math.pow(10, retryCount - 1)), retryMaxWaitMs);
+                log.warn("Retrying Elasticsearch request in {} ms. Attempt {} of {}", waitDurationMs, retryCount, retryAttemptLimit);
+                try {
+                    Thread.sleep(waitDurationMs);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(String.format("Thread interrupted while waiting for retry attempt %d of %d", retryCount, retryAttemptLimit), e);
+                }
+            }
+            try {
+                return delegate.performRequest(request);
+            } catch (ResponseException e) {
+                if (!retryOnErrorCodes.contains(e.getResponse().getStatusLine().getStatusCode())) {
+                    throw e;
+                }
+                lastException = e;
+            }
```

Comment:
it seems logic under the if statement `if (retryCount > 0)` needs to be executed on exception only. I would suggest moving the whole logic under that if statement under the `catch`. In this case you avoid having `if (retryCount > 0)`.
I.e. : 
```
} catch (ResponseException e) {
  if (!retryOnErrorCodes.contains(e.getResponse().getStatusLine().getStatusCode())) {
    throw e;
  }
  lastException = e;
  // put here that `sleep` logic.
}
```

Reply:
I was going to comment back that doing so would mean we'd do a wait when we'd end up breaking out of the loop anyways if we don't have that check. But I realized we could condense the loop's condition into the `if` statement and then get the best of both worlds at the trade off of doing the loop as a `while (true)`.

Done in 13e65bf87

Reply:
Thank you. Looks great!